src/main/java/org/opentripplanner/routing/core/RoutingContext.java

```diff
@@ -384,6 +384,13 @@ private void setServiceDays() {
                 addIfNotExists(this.serviceDays, new ServiceDay(graph, serviceDate, calendarService, agency.getId()));
                 addIfNotExists(this.serviceDays, new ServiceDay(graph, serviceDate.next(),
                         calendarService, agency.getId()));
+                if (opt.serviceDayLookout > 0) {
+                    ServiceDate sd = serviceDate;
+                    for (int i = 0; i < opt.serviceDayLookout; i++) {
```

Comment:
This is copying the original ServiceDate and then incrementing/decrementing it before adding it. It looks like this will double-add the day immediately after or before the original ServiceDate (since it's already been added on line 385 or 383 above). Rather than having duplicate logic to handle the default "one day on either side" case and the asymmetric "long lookout on one side of the start date" case, it would make more sense to me if the serviceDayLookout parameter just defaulted to 1 and the default (existing) behavior was just treated as a minimum case of service day lookout.

Reply:
I'm happy to make then new behavior the default.

Double-adding won't cause correctness problems because of the use of `addIfNotExists`. If there's a performance penalty, the larger issue with double-adding is that we loop over all feeds and all agencies, but any ServiceDay which is created for a timezone which already exists is redundant and won't end up getting added to the list. agencyId is only used to look up the timezone:
https://github.com/opentripplanner/OpenTripPlanner/blob/4820e9cbc59c77ff832a492b9f1f497bf9834feb/src/main/java/org/opentripplanner/routing/core/ServiceDay.java#L42-L47

Shouldn't be too hard to refactor this.

Reply:
Thanks for the clarification. I haven't looked at this code for a while and was scratching my head over the agency loop (which is apparently mostly to deal with multiple time zones, even though it's frequently said elsewhere that OTP assumes one time zone per graph), but now I realize there's a lot of de-duplication going on in the list of service dates. An additional performance concern is that these dates are being added to a list rather than a set, so each check for duplicates is linear-time. But I think this list is usually so short that the linear-time solution is just as fast.

So the existing implementation looks fine, it's just a bit strange to add one extra day, then re-iterate to add some of the same days. It does very simply handle the fact that you want one day before and N days after though. I'm fine with keeping the existing implementation or trying to generalize it so that even the default is a 1-day lookout, whichever you prefer.